### ISSUE TITLE
Do we need to implement a "send events you have" for desktop applications?

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/mozilla-vpn-client#4222]
After the work to move glean to the android sdk and implement a “send pings before you leave” we are seeing a phenomenon that we haven’t seen before. It’s most evident in the combination of the deletion requests view - [https://mozilla.cloud.looker.com/dashboards/429](https://mozilla.cloud.looker.com/dashboards/429) and the conversion funnel view [https://mozilla.cloud.looker.com/dashboards/472](https://mozilla.cloud.looker.com/dashboards/472).

If we filter the views to look at just Android and version 2.9.0 there is a marked discrepancy between the number of clients submitting data and the number of deletion requests. We can also see this discrepancy in the [conversion funnel view](https://mozilla.cloud.looker.com/dashboards/472). Because clients have to complete the mobile conversion funnel in order to be able to opt-out of data we see a lot of data from clients who open the app but don’t complete a subscription.

The scale of this new data coming in from android is surprising. Should we be expecting this on desktop platforms too but aren’t seeing it because we don’t send events if people bail from the client early? Or are desktop users more “qualified” users because they probably came through the desktop subscription flow and so are more likely to open the app with a subscription in hand?

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/VPN-2704)
┆Reporter: Sarah Bird

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

{quote}Because clients have to complete the mobile conversion funnel in order to be able to opt-out of data we see a lot of data from clients who open the app but don’t complete a subscription.{quote}

This sounds to me like a bug. According to the Mozilla data collection guidelines ( https://wiki.mozilla.org/Data_Collection ) users must have a way of turning off default on data collection. I will file a separate ticket to discuss this.

{quote}Should we be expecting this on desktop platforms too but aren’t seeing it because we don’t send events if people bail from the client early?{quote}

I guess there is no way to know unless we have the same ping scheduling on all platforms, so IMO yes, we should send a ping before shutdown on Desktop as well. Hm, but don’t we do that already? …I check… 

Nope, we don’t! I thought we did because we do shutdown Glean when we are about to quit ( https://github.com/mozilla-mobile/mozilla-vpn-client/blob/9e242d2374c4c665e5f89bab3c8738fe08295ac8/src/ui/main.qml#L347 ). Implementing this would be as simple as adding a “send ping” call right before that shutdown call. We also need wait for shutdown to resolve before quitting, however that is already a problem with the implementation right now.

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

"Implementing this would be as simple as adding a “send ping” call right before that shutdown call."

Even if it doesn't solve all the issues you've identified, would this really be a simple PR. If yes, we should do it.

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Yeap, it would be a really simple PR https://github.com/mozilla-mobile/mozilla-vpn-client/pull/4225 ( https://github.com/mozilla-mobile/mozilla-vpn-client/pull/4225|smart-link ). The only thing that is worrying is that we do not wait on the shutdown method, so there is no guarantee it completed before quitting and so no guarantee that last ping is sent.

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Adding this to the migration epic because it will be easier to do once that is implemented.

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

This is in review as part of https://mozilla-hub.atlassian.net/browse/VPN-3055 ( https://mozilla-hub.atlassian.net/browse/VPN-3055|smart-link ). Specifically: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5148/commits/fa82e33ff90142602a07f3125c815b45ca138fcd#diff-705b181dfea5ef23cadef0ed0e133fdf1192f2d41d0ba9435e4b7f5ecd8f8cfcR563 ( https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5148/commits/fa82e33ff90142602a07f3125c815b45ca138fcd#diff-705b181dfea5ef23cadef0ed0e133fdf1192f2d41d0ba9435e4b7f5ecd8f8cfcR563|smart-link )
